frb/analysis/kcwi.py

```diff
@@ -113,358 +66,473 @@ def silence_warnings(warncategory):
-        hdulist.writeto(specfile_name, overwrite=True)
-
-        # Append spectrum to list
-        speclist[idx] = spec.data
-        varspeclist[idx] = varspec.data
-
-    if marzfile:
-        _make_marz(cube, speclist, varspeclist, objects, outdir+marzfile, tovac=tovac)
-    return speclist, varspeclist, wave
+        # if kind is max:
+        # if kind is something_else:
+        masked_var = self.varcube.subcube_from_mask(mask_arr)
+        varspec = masked_var.mean(axis = (1,2), how = how)/np.sum(mask_arr) # Var of the mean = Sum of var/npix^2 = Mean of var/npix assuming each pix is uncorelated (not true, really).
+        return spec, varspec
+
+    def spec_from_ellipse(self,
+                        x0 = 0., y0 = 0., a = 1.,
+                        b = 1., theta = 0.):
+        
+        """
+        Get the spectrum within an elliptical region
+        Args:
+            cube (Spectral Cube): A datacube object
+            x0, y0 (float, optional): Centroid of ellipse
+            a, b (float, optional): semi-major and semi-minor axes
+            theta (float, optional): rotation angle of the semi-major
+                axis from the positive x axis.
+        Returns:
+            spec (OneDSpectrum): The extracted spectrum.
+            var (OneDSpectrum): Variance in spectrum.
+                Only returned if varcube is supplied.
+        """
+        #TODO: Use photutils aperture object for this.
+        # Create 2D mask first
+        ellipse = EllipticalAperture((x0, y0), a, b, theta)
+        # Create a mask from the ellipse
+        aper_mask = ellipse.to_mask(method='center').to_image(self.cube.shape[1:]).astype(bool)
+        
+        return self.spec_from_spatial_mask(aper_mask)
+
+
+    def _make_marz(self, speclist, varspeclist, objects,  wave=None, marzfile="marzfile.fits", tovac = True):
+        """
+        Helper function to create a MARZ input file
+        Args:
+            cube (Spectral cube): Datacube
+            speclist (list): list of spectra. i.e.
+                Spectral cube Projections of same
+                shape as cube.spectral_axis.
+            varspeclist (list): list of spectral
+                variances (same object class and
+                shape as speclist elements)
+            objects (astropy Table): Table of
+                objects detected using SourceCatalog.
+            wave (ndarray, optional): Wavelength array.
+                If None, the cube.spectral_axis
+                is used.
+            marzfile (str, optional): Name of
+                output MARZ fits file.
+            tovac (bool, optional): Convert wavelengths
+                to vacuum?
+        """
+        # TODO: Actually compute sky background
+        nobjs = len(objects)
+        wcsinfo = self.cube.wcs.celestial
+
+        sky = np.zeros_like(speclist)
+
+        # Convert wavelengths from air to vacuum
+        if wave is None:
+            wave = self.cube.spectral_axis.value
+        if tovac:
```

Comment:
do you need to save the `tovac` keyword to the output file such that you don't lose track of which marz files have `tovac` vs not?

Reply:
Good point. I need to look into MARZ docs if there's a consistent way to do this.

Reply:
So MARZ just looks for a header keyword. I've now switched to providing that instead of doing the conversion myself. Helps keep track explicitly and prevents multiple conversions.